src/Chainweb/Pact/PactService.hs

```diff
@@ -407,13 +421,54 @@ readAccountGuard
 readAccountGuard pdb account
     = fmap ssnd <$> readCoinAccount pdb account
 
+-- | Rewards table mapping 3-month periods to their rewards
+-- according to the calculated exponential decay over 120 year period
+--
+readRewards
+    :: HasChainGraph v
+    => v -> IO (HashMap BlockHeight P.ParsedDecimal)
+readRewards v = do
+    rs <- LBS.readFile "rewards/miner_rewards.csv"
+    case CSV.decode CSV.NoHeader rs of
+      Left e -> internalError
+        $ "cannot construct miner reward map: "
+        <> sshow e
+      Right vs -> return
+        $ HM.fromList
+        . V.toList
+        . V.map formatRow
+        $ vs
+  where
+    formatRow :: (Word64, Double) -> (BlockHeight, P.ParsedDecimal)
+    formatRow (!a,!b) =
+      let
+        !n = v ^. chainGraph . to (int . size)
+        !m = fromRational $ toRational b
+      in (BlockHeight a, P.ParsedDecimal $ m / n)
+
+-- | Calculate miner reward. We want this to error hard in the case where
+-- block times have finally exceeded the 120-year range. Rewards are calculated
+-- in 500k steps
+--
+minerReward
+    :: forall cas
+    . BlockHeight -> PactServiceM cas P.ParsedDecimal
+minerReward bh = do
+    m <- view $ psMinerRewards . at (roundBy bh 500000)
+    case m of
+      Nothing -> internalError
+          $ "block height outside of admissible range: "
+          <> sshow bh
+      Just r -> return r
+{-# INLINABLE minerReward #-}
```

Comment:
In 120 years, we'll start seeing the `Nothing` case

Reply:
Is there some kind of closed form we could use for this at all further points in time?